File: ModernHotelSYS.java
import java.io.*;
import java.time.LocalDate;
import java.util.*;

class Room {
    private int roomNumber;
    private String roomType;
    private double pricePerNight;

    public Room(int roomNumber, String roomType, double pricePerNight) {
        this.roomNumber = roomNumber;
        this.roomType = roomType;
        this.pricePerNight = pricePerNight;
    }

    public int getRoomNumber() {
        return roomNumber;
    }

    public String getRoomType() {
        return roomType;
    }

    public double getPricePerNight() {
        return pricePerNight;
    }

    @Override
    public String toString() {
        return "Room " + roomNumber + " (" + roomType + ") - $" + pricePerNight + "/night";
    }
}

class Guest {
    private String name;
    private String contactInfo;

    public Guest(String name, String contactInfo) {
        this.name = name;
        this.contactInfo = contactInfo;
    }

    public String getName() {
        return name;
    }

    public String getContactInfo() {
        return contactInfo;
    }

    @Override
    public String toString() {
        return "Guest: " + name + ", Contact: " + contactInfo;
    }
}

class Reservation {
    private String reservationId;
    private Guest guest;
    private Room room;
    private LocalDate checkInDate;
    private LocalDate checkOutDate;

    public Reservation(String reservationId, Guest guest, Room room, LocalDate checkInDate, LocalDate checkOutDate) {
        this.reservationId = reservationId;
        this.guest = guest;
        this.room = room;
        this.checkInDate = checkInDate;
        this.checkOutDate = checkOutDate;
    }

    public String getReservationId() {
        return reservationId;
    }

    public Guest getGuest() {
        return guest;
    }

    public Room getRoom() {
        return room;
    }

    public LocalDate getCheckInDate() {
        return checkInDate;
    }

    public LocalDate getCheckOutDate() {
        return checkOutDate;
    }

    public double calculateTotalCost() {
        long nights = java.time.temporal.ChronoUnit.DAYS.between(checkInDate, checkOutDate);
        return nights * room.getPricePerNight();
    }

    @Override
    public String toString() {
        return "Reservation ID: " + reservationId + "\n"
                + "Guest: " + guest.getName() + "\n"
                + "Room: " + room.getRoomNumber() + " (" + room.getRoomType() + ")\n"
                + "Check-in: " + checkInDate + ", Check-out: " + checkOutDate + "\n"
                + "Total Cost: $" + calculateTotalCost();
    }
}

class Hotel {
    private String name;
    private String address;
    private List<Room> rooms = new ArrayList<>();
    private List<Reservation> reservations = new ArrayList<>();
    private int reservationCounter = 1;

    public Hotel(String name, String address) {
        this.name = name;
        this.address = address;
    }

    public void addRoom(Room room) {
        rooms.add(room);
    }

    public void autoAddRooms(int singles, int doubles, int suites) {
        for (int i = 1; i <= singles; i++)
            addRoom(new Room(100 + i, "Single", 100));
        for (int i = 1; i <= doubles; i++)
            addRoom(new Room(200 + i, "Double", 150));
        for (int i = 1; i <= suites; i++)
            addRoom(new Room(300 + i, "Suite", 300));
    }

    // ✅ New method: checks if a room is free for given dates
    public boolean isRoomAvailable(Room room, LocalDate checkIn, LocalDate checkOut) {
        for (Reservation r : reservations) {
            if (r.getRoom().getRoomNumber() == room.getRoomNumber()) {
                if (!(checkOut.isBefore(r.getCheckInDate()) || checkIn.isAfter(r.getCheckOutDate()))) {
                    return false;
                }
            }
        }
        return true;
    }

    public Room findAvailableRoom(String type, LocalDate checkIn, LocalDate checkOut) {
        for (Room r : rooms) {
            if (r.getRoomType().equalsIgnoreCase(type) && isRoomAvailable(r, checkIn, checkOut)) {
                return r;
            }
        }
        return null;
    }

    public Reservation makeReservation(Guest guest, String type, LocalDate checkIn, int nights) {
        LocalDate checkOut = checkIn.plusDays(nights);
        Room room = findAvailableRoom(type, checkIn, checkOut);
        if (room != null) {
            Reservation reservation = new Reservation("RES-" + (reservationCounter++), guest, room, checkIn, checkOut);
            reservations.add(reservation);
            saveReservationsToFile();
            return reservation;
        }
        return null;
    }

    public void cancelReservation(String reservationId) {
        Iterator<Reservation> it = reservations.iterator();
        while (it.hasNext()) {
            Reservation r = it.next();
            if (r.getReservationId().equals(reservationId)) {
                it.remove();
                saveReservationsToFile();
                System.out.println("Reservation " + reservationId + " canceled.");
                return;
            }
        }
        System.out.println("Reservation ID not found.");
    }

    public void listAvailableRooms(LocalDate checkIn, int nights, String type) {
        LocalDate checkOut = checkIn.plusDays(nights);
        boolean found = false;
        for (Room r : rooms) {
            if (r.getRoomType().equalsIgnoreCase(type) && isRoomAvailable(r, checkIn, checkOut)) {
                System.out.println(r);
                found = true;
            }
        }
        if (!found)
            System.out.println("No available rooms for the given date.");
    }

    public void listReservations() {
        if (reservations.isEmpty())
            System.out.println("No current reservations.");
        else
            for (Reservation r : reservations) {
                System.out.println(r);
                System.out.println("----------------------------------------");
            }
    }

    public void saveReservationsToFile() {
        try (PrintWriter writer = new PrintWriter(new File("reservations.csv"))) {
            writer.println("Reservation ID,Guest Name,Room Number,Room Type,Check-in Date,Check-out Date,Total Cost");
            for (Reservation r : reservations) {
                writer.println(r.getReservationId() + "," + r.getGuest().getName() + "," +
                        r.getRoom().getRoomNumber() + "," + r.getRoom().getRoomType() + "," +
                        r.getCheckInDate() + "," + r.getCheckOutDate() + "," + r.calculateTotalCost());
            }
        } catch (IOException e) {
            System.out.println("Error saving reservations: " + e.getMessage());
        }
    }

    public void loadReservationsFromFile() {
        try (BufferedReader reader = new BufferedReader(new FileReader("reservations.csv"))) {
            String line = reader.readLine();
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length < 7)
                    continue;
                Guest guest = new Guest(parts[1], "N/A");
                Room room = findAvailableRoom(parts[3], LocalDate.parse(parts[4]), LocalDate.parse(parts[5]));
                if (room != null) {
                    reservations.add(new Reservation(parts[0], guest, room, LocalDate.parse(parts[4]),
                            LocalDate.parse(parts[5])));
                }
            }
        } catch (IOException e) {
            System.out.println("Error loading reservations: " + e.getMessage());
        }
    }
}

public class ModernHotelSYS {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Hotel hotel = new Hotel("Modern Hotel", "123 Main Street");
        hotel.autoAddRooms(10, 10, 10);
        hotel.loadReservationsFromFile();

        while (true) {
            System.out.println("\nWelcome to Modern Hotel Reservation System");
            System.out.println("1. Make a Reservation");
            System.out.println("2. View Current Reservations");
            System.out.println("3. Cancel a Reservation");
            System.out.println("4. Exit");
            System.out.print("Enter your choice: ");
            String choice = scanner.nextLine();

            switch (choice) {
                case "1":
                    System.out.print("Enter guest name: ");
                    String name = scanner.nextLine();
                    System.out.print("Enter guest contact info: ");
                    String contact = scanner.nextLine();
                    Guest guest = new Guest(name, contact);

                    System.out.print("Enter room type (Single/Double/Suite): ");
                    String type = scanner.nextLine();

                    LocalDate checkIn = null;
                    while (checkIn == null) {
                        System.out.println("Select Check-in Date:");
                        System.out.println("1. Today (" + LocalDate.now() + ")");
                        System.out.println("2. Tomorrow (" + LocalDate.now().plusDays(1) + ")");
                        System.out.println("3. Enter custom date");
                        System.out.print("Choice: ");
                        String dateChoice = scanner.nextLine();

                        switch (dateChoice) {
                            case "1":
                                checkIn = LocalDate.now();
                                break;
                            case "2":
                                checkIn = LocalDate.now().plusDays(1);
                                break;
                            case "3":
                                System.out.print("Enter check-in date (YYYY-MM-DD): ");
                                try {
                                    checkIn = LocalDate.parse(scanner.nextLine());
                                } catch (Exception e) {
                                    System.out.println("Invalid format. Try again.");
                                }
                                break;
                            default:
                                System.out.println("Invalid choice. Try again.");
                        }
                    }


                    System.out.print("How many nights will the guest stay? ");
                    int nights = Integer.parseInt(scanner.nextLine());

                    hotel.listAvailableRooms(checkIn, nights, type);
                    Reservation res = hotel.makeReservation(guest, type, checkIn, nights);
                    if (res != null) {
                        System.out.println("Reservation successful!");
                        System.out.println(res);
                    } else {
                        System.out.println("No available room for that period.");
                    }
                    break;
                case "2":
                    hotel.listReservations();
                    break;
                case "3":
                    System.out.print("Enter reservation ID to cancel: ");
                    hotel.cancelReservation(scanner.nextLine());
                    break;
                case "4":
                    System.out.println("Thank you for using the system. Goodbye!");
                    return;
                default:
                    System.out.println("Invalid option.");
            }
        }
    }
}

File: Modern_HotelSYS.py
import csv
import datetime
import uuid

# Define the Room class
class Room:
    def __init__(self, room_number, room_type, price_per_night, is_available=True):
        self.room_number = room_number
        self.room_type = room_type
        self.price_per_night = price_per_night
        self.is_available = is_available

    def __str__(self):
        return f"Room {self.room_number} ({self.room_type}) - ${self.price_per_night}/night"

# Define the Guest class
class Guest:
    def __init__(self, name, contact_info):
        self.name = name
        self.contact_info = contact_info

    def __str__(self):
        return f"Guest: {self.name}, Contact: {self.contact_info}"

# Define the Reservation class
class Reservation:
    def __init__(self, guest, room, check_in_date, check_out_date):
        self.reservation_id = str(uuid.uuid4())  # Generate a unique reservation ID
        self.guest = guest
        self.room = room
        self.check_in_date = check_in_date
        self.check_out_date = check_out_date

    def calculate_total_cost(self):
        num_nights = (self.check_out_date - self.check_in_date).days
        return num_nights * self.room.price_per_night

    def __str__(self):
        return (f"Reservation ID: {self.reservation_id}\n"
                f"Guest: {self.guest.name}\n"
                f"Room: {self.room.room_number} ({self.room.room_type})\n"
                f"Check-in: {self.check_in_date}, Check-out: {self.check_out_date}\n"
                f"Total Cost: ${self.calculate_total_cost()}")

# Define the Hotel class
class Hotel:
    def __init__(self, name, address):
        self.name = name
        self.address = address
        self.rooms = []
        self.reservations = []

    def add_room(self, room):
        self.rooms.append(room)

    def auto_add_rooms(self, num_singles, num_doubles, num_suites):
        for i in range(1, num_singles + 1):
            self.add_room(Room(100 + i, "Single", 100))
        for i in range(1, num_doubles + 1):
            self.add_room(Room(200 + i, "Double", 150))
        for i in range(1, num_suites + 1):
            self.add_room(Room(300 + i, "Suite", 300))

    def find_available_room(self, room_type):
        for room in self.rooms:
            if room.room_type.lower() == room_type.lower() and room.is_available:
                return room
        return None

    def list_available_rooms(self):
        available_rooms = [room for room in self.rooms if room.is_available]
        if available_rooms:
            print("Available Rooms:")
            for room in available_rooms:
                print(room)
        else:
            print("No rooms available.")

    def list_reservations(self):
        if self.reservations:
            print("\nCurrent Reservations:")
            for reservation in self.reservations:
                print(reservation)
                print("-" * 40)
        else:
            print("No current reservations.")

    def make_reservation(self, guest, room_type, check_in_date, check_out_date):
        if check_in_date >= check_out_date:
            print("Error: Check-in date must be before check-out date.")
            return None

        room = self.find_available_room(room_type)
        if room:
            room.is_available = False
            reservation = Reservation(guest, room, check_in_date, check_out_date)
            self.reservations.append(reservation)
            self.save_reservations_to_file()  # Save reservations to file
            return reservation
        else:
            print("No available rooms for the selected type.")
            return None

    def cancel_reservation(self, reservation_id):
        for reservation in self.reservations:
            if reservation.reservation_id == reservation_id:
                confirm = input(f"Are you sure you want to cancel reservation {reservation_id}? (yes/no): ").lower()
                if confirm != "yes":
                    print("Cancellation aborted.")
                    return

                reservation.room.is_available = True
                self.reservations.remove(reservation)
                self.save_reservations_to_file()  # Save reservations to file
                print(f"Reservation {reservation_id} has been canceled.")
                return
        print(f"Reservation ID {reservation_id} not found.")

    def __str__(self):
        return f"{self.name}, located at {self.address}"

    def save_reservations_to_file(self, filename="reservations.csv"):
        """Save all reservations to a CSV file."""
        with open(filename, mode="w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["Reservation ID", "Guest Name", "Room Number", "Room Type", "Check-in Date", "Check-out Date", "Total Cost"])
            for reservation in self.reservations:
                writer.writerow([
                    reservation.reservation_id,
                    reservation.guest.name,
                    reservation.room.room_number,
                    reservation.room.room_type,
                    reservation.check_in_date,
                    reservation.check_out_date,
                    reservation.calculate_total_cost()
                ])
        print(f"Reservations saved to {filename}.")

    def load_reservations_from_file(self, filename="reservations.csv"):
        """Load reservations from a CSV file (optional feature)."""
        try:
            with open(filename, mode="r") as file:
                reader = csv.DictReader(file)
                for row in reader:
                    guest = Guest(row["Guest Name"], "N/A")
                    room = self.find_available_room(row["Room Type"])
                    if room:
                        room.is_available = False
                        check_in_date = datetime.datetime.strptime(row["Check-in Date"], "%Y-%m-%d").date()
                        check_out_date = datetime.datetime.strptime(row["Check-out Date"], "%Y-%m-%d").date()
                        reservation = Reservation(guest, room, check_in_date, check_out_date)
                        self.reservations.append(reservation)
            print(f"Reservations loaded from {filename}.")
        except FileNotFoundError:
            print(f"No existing reservation file found: {filename}.")


def main():
    hotel = Hotel("Modern Hotel", "123 Main Street")
    hotel.auto_add_rooms(10, 10, 10)
    hotel.load_reservations_from_file()

    while True:
        print("\nWelcome to Modern Hotel Reservation System")
        print("1. List Available Rooms")
        print("2. Make a Reservation")
        print("3. Cancel a Reservation")
        print("4. View Current Reservations")
        print("5. Exit")
        choice = input("Enter your choice: ")

        if choice == "1":
            hotel.list_available_rooms()
        elif choice == "2":
            name = input("Enter guest name: ")
            contact_info = input("Enter guest contact info: ")
            guest = Guest(name, contact_info)

            valid_room_types = ["Single", "Double", "Suite"]
            room_type_input = input("Enter room type (Single/Double/Suite): ").title()
            if room_type_input not in valid_room_types:
                print("Invalid room type. Please enter Single, Double, or Suite.")
                continue

            try:
                check_in_str = input("Enter check-in date (MM/DD): ")
                check_out_str = input("Enter check-out date (MM/DD): ")
                current_year = datetime.date.today().year
                check_in_date = datetime.datetime.strptime(f"{check_in_str}/{current_year}", "%m/%d/%Y").date()
                check_out_date = datetime.datetime.strptime(f"{check_out_str}/{current_year}", "%m/%d/%Y").date()
            except ValueError:
                print("Invalid date format. Please use MM/DD.")
                continue

            reservation = hotel.make_reservation(guest, room_type_input, check_in_date, check_out_date)
            if reservation:
                print("Reservation successful!")
                print(reservation)
        elif choice == "3":
            reservation_id = input("Enter reservation ID to cancel: ")
            hotel.cancel_reservation(reservation_id)
        elif choice == "4":
            hotel.list_reservations()
        elif choice == "5":
            print("Thank you for using Modern Hotel Reservation System. Goodbye!")
            break
        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()
